src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -154,195 +155,260 @@ protected override void DetachCore()
-                            {
-                                Network network = peer.Network;
-                                ChainedBlock consensusTip = this.chainState.ConsensusTip;
-                                if ((network != null) && (consensusTip != null))
-                                {
-                                    ChainedBlock fork = this.pendingTip.FindFork(consensusTip);
-                                    if ((fork != null) && (fork != consensusTip))
-                                    {
-                                        int reorgLength = consensusTip.Height - fork.Height;
-                                        if (reorgLength > maxReorgLength)
-                                        {
-                                            this.logger.LogTrace("Reorganization of length {0} prevented, maximal reorganization length is {1}, consensus tip is '{2}'.", reorgLength, maxReorgLength, consensusTip);
-                                            this.InvalidHeaderReceived = true;
-                                            reorgPrevented = true;
-                                        }
-                                        else this.logger.LogTrace("Reorganization of length {0} accepted, consensus tip is '{1}'.", reorgLength, consensusTip);
-                                    }
-                                }
-                            }
+                    if (prev == null)
+                    {
+                        this.logger.LogTrace("Previous header of the new header '{0}' was not found on our chain either.", header);
 
-                            // Switch to better chain.
-                            if (!reorgPrevented)
-                            {
-                                this.logger.LogTrace("New chain tip '{0}' selected, chain work is '{1}'.", this.pendingTip, this.pendingTip.ChainWork);
-                                this.Chain.SetTip(this.pendingTip);
-                            }
-                        }
+                        // If we can't connect the header we received from the peer, we might be on completely different chain or
+                        // a reorg happened recently. If we ignored it, we would have invalid view of the peer and the propagation
+                        // of blocks would not work well. So we ask the peer for headers using "getheaders" message.
+                        // Enforce a sync.
+                        doTrySync = true;
+                        break;
+                    }
+
+                    // Now we know the previous block header and thus we can connect the new header.
+                }
+
+                tip = new ChainedBlock(header, header.GetHash(peer.Network.NetworkOptions), prev);
+                bool validated = this.Chain.GetBlock(tip.HashBlock) != null || tip.Validate(peer.Network);
+                validated &= !this.chainState.IsMarkedInvalid(tip.HashBlock);
```

Comment:
Not a big fan of &= for boolean operations. It looked weird to me so I googled and it does seem valid (it's in C# case example) so I think it's ok but I personally would prefer putting that condition into a variable and adding an && at the end of the main condition.

Reply:
old code